### ISSUE TITLE
Move all platform folders in a `platforms` folder

### ISSUE THREAD

**Issue by provokateurin** [flutter/flutter#34536]
## Use case

As flutter now supports a lot of different platforms, it would be nice to move all platform folders in a platforms folder.
It would be a breaking change, but it could be fixed very easily manually of even by a script to  move these folders.

## Proposal
Before:
```
app/
    lib/
    android/
    ios/
    linux/
    macos
    windows/
    web/
```
After:
```
app/
    lib/
    platforms/
        android/
        ios/
        linux/
        macos
        windows/
        web/
```

**Comment by NBTX** [comment]
I feel like this is more logical anyway - especially for people exploring flutter projects.

**Comment by stuartmorgan** [comment]
Can you explain what problem this would solve? I can see the argument that all else being equal the structure you proposed could be cleaner, but changing now would have a huge cost in project churn, tooling, confusion in the established user base, etc.

**Comment by jld3103** [comment]
I know what you mean. Currently I think it's worth it, but it could be a better structure if flutter supports a lot more platforms in the future (?).

**Comment by jonahwilliams** [comment]
I don't really see the value in placing platform code in another subfolder

**Comment by stuartmorgan** [comment]
Again, could you describe the problem you are trying to solve with this suggestion, specifically?

**Comment by jld3103** [comment]
Ok. I'm developing an app that supports all these platforms and and is in the same folder as the server and another part. If you then open some of these folders it get's really messy. The only real advantage is, that it would be less messy. No more.

**Comment by lazylazyllama** [comment]
Why was this closed? I think it would be a huge improvement. I'm very annoyed by the current structure aswell and it could be easily done without confusing people by just having the new structure for new projects. So old projects could stay as they are and new projects will have the new structure. I don't think that this would confuse people too much ^^

**Comment by stuartmorgan** [comment]
> It should be possible to answer questions, just looking on folders

This is already possible, it's just slightly harder than it would be with subfolders.

As I mentioned above, this isn't a decision being made in vacuum for green-field development:

>> changing now would have a huge cost in project churn, tooling, confusion in the established user base, etc.

How would you propose handling all of that?
- Will all projects be force-migrated, or will `flutter` need to support both structures forever?
  - If the former:
    - What happens to projects that already happen to have a `platforms` directory?
    - Will we force-modify everyone's `.gitignore` files? Because sufficiently old projects have all their platform-specific ignores in a singe top-level `.gitignore`, so if we don't the git state of old projects will be suddenly wrong (but if we do, we have to auto-modify a file that has potentially arbitrary user changes).
    - What, if anything, will be done about the inconsistency with plugins, which also have all of these same directories, but can't be force-migrated?
    - How will we inform users that their project have changed out from under them?
  - If the latter:
    - What happens when someone tries to add a platform to an existing project with `flutter create`? Will they get a hybrid, or will that also need to support both structures forever?
    - Will all third-party tools/scripts that interact with the structure support both, or will there be a confusing ecosystem where some tools work on all project, while some only work on both?
    - Will we run twice as many project integration tests forever to make sure that everything still works in both configurations? Will people remember to test both configurations with new features?
- What would the plan be for updating all of our first-party documentation, which refers to the old paths everywhere?
- What happens to all of the third-party tooling out there that expects the current structure (e.g., the packages that streamline updating app icons or names across all platforms)?
- What is the cost in terms of user confusion for all the third-party content (courses, tutorial, SO answers, etc.) that may or may not ever be updated? Is there a plan to mitigate that somehow?

That's not every a comprehensive list, it's just off the top of my head.

**Comment by polina-c** [comment]
Yes, it is a big deal and folder reorg should be postponed till next very breaking change in Flutter, when we face all these issues anyway.

For now it would it be possible to sort the folders by type and color code them in VSCode?

**Comment by stuartmorgan** [comment]
> folder reorg should be postponed till next very breaking change in Flutter, when we face all these issues anyway.

We have breaking changes in essentially every Flutter release, but most of them aren't like this. I don't know that we have any precedent for breaking the *structure* of a Flutter project.

**Comment by stuartmorgan** [comment]
It looks like this was essentially re-filed at some point as https://github.com/flutter/flutter/issues/71400; closing as a duplicate of that since there's more voting and discussion there.
